velox/functions/lib/IsNull.cpp

```diff
@@ -32,47 +32,44 @@ class IsNullFunction : public exec::VectorFunction {
       const TypePtr& /* outputType */,
       exec::EvalCtx* context,
       VectorPtr* result) const override {
-    BaseVector* arg = args[0].get();
-    if (rows.isAllSelected()) {
-      if (!arg->mayHaveNulls()) {
-        // no nulls
-        *result =
-            BaseVector::createConstant(IsNotNULL, rows.end(), context->pool());
-        return;
-      }
+    auto* arg = args[0].get();
 
-      context->ensureWritable(rows, BOOLEAN(), *result);
-      FlatVector<bool>* flatResult = (*result)->asFlatVector<bool>();
-      flatResult->clearNulls(rows);
-      flatResult->mutableRawValues<int64_t>();
-      auto rawNulls = arg->flatRawNulls(rows);
-      memcpy(
-          flatResult->mutableRawValues<int64_t>(),
-          rawNulls,
-          bits::nbytes(rows.end()));
-      if constexpr (!IsNotNULL) {
-        bits::negate(flatResult->mutableRawValues<char>(), rows.end());
-      }
+    // Input can be flat or constant.
+    if (arg->isConstantEncoding()) {
+      bool isNull = arg->isNullAt(rows.begin());
+      auto localResult = BaseVector::createConstant(
+          IsNotNULL ? !isNull : isNull, rows.size(), context->pool());
+      context->moveOrCopyResult(localResult, rows, *result);
       return;
     }
 
-    context->ensureWritable(rows, BOOLEAN(), *result);
-    FlatVector<bool>* flatResult = (*result)->asFlatVector<bool>();
     if (!arg->mayHaveNulls()) {
-      rows.applyToSelected(
-          [&](vector_size_t i) { flatResult->set(i, IsNotNULL); });
+      // No nulls.
+      auto localResult = BaseVector::createConstant(
+          IsNotNULL ? true : false, rows.size(), context->pool());
+      context->moveOrCopyResult(localResult, rows, *result);
+      return;
+    }
+
+    BufferPtr isNull;
+    if constexpr (IsNotNULL) {
+      isNull = arg->nulls();
     } else {
-      auto rawNulls = arg->flatRawNulls(rows);
-      if constexpr (!IsNotNULL) {
-        rows.applyToSelected([&](vector_size_t i) {
-          flatResult->set(i, bits::isBitNull(rawNulls, i));
-        });
-      } else {
-        rows.applyToSelected([&](vector_size_t i) {
-          flatResult->set(i, !bits::isBitNull(rawNulls, i));
-        });
-      }
+      isNull = AlignedBuffer::allocate<bool>(rows.size(), context->pool());
+      memcpy(
+          isNull->asMutable<int64_t>(),
+          arg->rawNulls(),
+          bits::nbytes(rows.end()));
+      bits::negate(isNull->asMutable<char>(), rows.end());
```

Comment:
We no longer check the selectivity?

Reply:
It is not needed. Most of the time 'result' doesn't have any data that requires preserving, hence, moveOrCopyResult will simply move the localResult. In rare cases when is_null if used under a switch or if, moveOrCopyResult will copy the data as needed.

Reply:
I think in the old version, when selectivity is set and result is reused, we do not allocate new buffer but only overwrite the selected rows in the result.  In the new version we do the allocation and compute the full range regardless of the selectivity.

This case is very rare though, only happens with a low selectivity, negated nullability, and reuse of the result.

Reply:
That's correct.